packages/evm/src/opcodes/functions.ts

```diff
@@ -796,24 +796,32 @@ export const handlers: Map<number, OpHandler> = new Map([
       runState.programCounter = Number(dest)
     },
   ],
-  // 0x5e: JUMPSUB
+  // 0x5e: JUMPSUB (2315) / MCOPY (5656)
   [
     0x5e,
-    function (runState) {
-      const dest = runState.stack.pop()
+    function (runState, common) {
+      if (common.isActivatedEIP(2315)) {
+        // JUMPSUB
+        const dest = runState.stack.pop()
 
-      if (dest > runState.interpreter.getCodeSize()) {
-        trap(ERROR.INVALID_JUMPSUB + ' at ' + describeLocation(runState))
-      }
+        if (dest > runState.interpreter.getCodeSize()) {
+          trap(ERROR.INVALID_JUMPSUB + ' at ' + describeLocation(runState))
+        }
 
-      const destNum = Number(dest)
+        const destNum = Number(dest)
 
-      if (!jumpSubIsValid(runState, destNum)) {
-        trap(ERROR.INVALID_JUMPSUB + ' at ' + describeLocation(runState))
-      }
+        if (!jumpSubIsValid(runState, destNum)) {
+          trap(ERROR.INVALID_JUMPSUB + ' at ' + describeLocation(runState))
+        }
 
-      runState.returnStack.push(BigInt(runState.programCounter))
-      runState.programCounter = destNum + 1
+        runState.returnStack.push(BigInt(runState.programCounter))
+        runState.programCounter = destNum + 1
+      } else if (common.isActivatedEIP(5656)) {
+        // MCOPY
+        const [dst, src, length] = runState.stack.popN(3)
+        const data = runState.memory.read(Number(src), Number(length), true)
+        runState.memory.write(Number(dst), Number(length), data)
```

Comment:
An example test case wasn't provided either. Maybe we could add one for it as well.

Reply:
Thanks for this comment! Note that in `write`, we also call `extend` in memory. I am rather sure (and I think I once introduced this) is that this explicit error check is added as a precaution. However, I doubt this can ever happen.

( see `memory.ts` line 55: `this.extend(offset, size)`, this extends memory prior to writing)

(The unit tests absolutely do not cover all the cases. I have introduced them for completeness since these are in the EIP. However, EIP authors have written tests: https://github.com/ethereum/tests/pull/1229 which we pass)